src/core/ext/filters/client_channel/resolver/polling_resolver.cc

```diff
@@ -0,0 +1,201 @@
+//
+
+#include <grpc/support/port_platform.h>
+
+#include "src/core/ext/filters/client_channel/resolver/polling_resolver.h"
+
+#include "absl/strings/strip.h"
+
+#include "src/core/lib/backoff/backoff.h"
+#include "src/core/lib/channel/channel_args.h"
+#include "src/core/lib/iomgr/pollset_set.h"
+#include "src/core/lib/iomgr/timer.h"
+#include "src/core/lib/iomgr/work_serializer.h"
+
+namespace grpc_core {
+
+PollingResolver::PollingResolver(ResolverArgs args,
+                                 const grpc_channel_args* channel_args,
+                                 Duration min_time_between_resolutions,
+                                 BackOff::Options backoff_options,
+                                 TraceFlag* tracer)
+    : authority_(args.uri.authority()),
+      name_to_resolve_(absl::StripPrefix(args.uri.path(), "/")),
+      channel_args_(grpc_channel_args_copy(channel_args)),
+      work_serializer_(std::move(args.work_serializer)),
+      result_handler_(std::move(args.result_handler)),
+      tracer_(tracer),
+      interested_parties_(args.pollset_set),
+      min_time_between_resolutions_(min_time_between_resolutions),
+      backoff_(backoff_options) {
+  if (GPR_UNLIKELY(tracer_ != nullptr && tracer_->enabled())) {
+    gpr_log(GPR_INFO, "[polling resolver %p] created", this);
+  }
+}
+
+PollingResolver::~PollingResolver() {
+  if (GPR_UNLIKELY(tracer_ != nullptr && tracer_->enabled())) {
+    gpr_log(GPR_INFO, "[polling resolver %p] destroying", this);
+  }
+  grpc_channel_args_destroy(channel_args_);
+}
+
+void PollingResolver::StartLocked() { MaybeStartResolvingLocked(); }
+
+void PollingResolver::RequestReresolutionLocked() {
+  if (request_ == nullptr) {
+    MaybeStartResolvingLocked();
+  }
+}
+
+void PollingResolver::ResetBackoffLocked() {
+  if (have_next_resolution_timer_) {
+    grpc_timer_cancel(&next_resolution_timer_);
+  }
+  backoff_.Reset();
+}
+
+void PollingResolver::ShutdownLocked() {
+  if (GPR_UNLIKELY(tracer_ != nullptr && tracer_->enabled())) {
+    gpr_log(GPR_INFO, "[polling resolver %p] shutting down", this);
+  }
+  shutdown_ = true;
+  if (have_next_resolution_timer_) {
+    grpc_timer_cancel(&next_resolution_timer_);
+  }
+  request_.reset();
+}
+
+void PollingResolver::OnNextResolution(void* arg, grpc_error_handle error) {
+  auto* self = static_cast<PollingResolver*>(arg);
+  (void)GRPC_ERROR_REF(error);  // ref owned by lambda
+  self->work_serializer_->Run(
+      [self, error]() { self->OnNextResolutionLocked(error); }, DEBUG_LOCATION);
+}
+
+void PollingResolver::OnNextResolutionLocked(grpc_error_handle error) {
+  if (GPR_UNLIKELY(tracer_ != nullptr && tracer_->enabled())) {
+    gpr_log(GPR_INFO,
+            "[polling resolver %p] re-resolution timer fired: error=\"%s\", "
+            "request_=%p",
+            this, grpc_error_std_string(error).c_str(), request_.get());
+  }
+  have_next_resolution_timer_ = false;
+  if (error == GRPC_ERROR_NONE && request_ == nullptr) {
```

Comment:
Shouldn't it be a impossible for `request_` to not be nullptr at this point?

Also, I think we should check `shutdown_` here, and not start a new resolution if shutdown has started. Otherwise , I believe we could start a request that might never be explicitly cancelled.

Reply:
Related to above changes, let's log shutdown_ state in the above trace.